### PR TITLE
Custom notification sounds for rooms

### PR DESCRIPTION
**Pull request by Half-Shot**
Fixes https://github.com/vector-im/riot-web/issues/2105

image

This PR mainly exists so my alert noise for infrastructure breaking isn't the same noise as friendly PMs, which is probably heathy for my emotional state.

This is a very barebones feature for setting custom sounds for notifications within Riot. Riot reads for a room's "m.notification.sound" event and tries to load a mxcurl into a new audio element if ones doesn't already exist. 

It does not support multiple sources for a sound.

Epic for this feature can be found at: https://github.com/vector-im/riot-web/issues/9687
Riot-web labs.md PR can be found at: https://github.com/vector-im/riot-web/pull/9688

### CODE REVIEW

**Comment by turt2live** [comment]
In general, settings should use the SettingsStore instead of going direct.

**Comment by Half-Shot** [comment]
@turt2live I forgot that existed. Updated.

File: src/Notifier.js [hunk 1]
```diff
@@ -96,11 +96,52 @@ const Notifier = {
...
+            url: MatrixClientPeg.get().mxcUrlToHttp(content.url),
+            name: content.name,
+            type: content.type,
+            size: content.size,
+        };
+    },
+
+    _playAudioNotification: function(ev, room) {
+        this.getSoundForRoom(room.roomId).then((sound) => {
+            console.log(`Got sound ${sound.name || "default"} for ${room.roomId}`);
```
Comment (dnaf) [inline src/Notifier.js]: This line causes an error when `sound` is undefined.
```suggestion
            console.log(`Got sound ${sound ? sound.name : "default"} for ${room.roomId}`);
```

**Comment by Half-Shot** [comment]
I've updated the description with a list of issues and corners cut when doing the PR for transparency. I suppose we want to fix some of them before merging but I believe some are quite minor.

File: src/settings/handlers/AccountSettingsHandler.js [hunk 2]
```diff
@@ -73,6 +73,10 @@ export default class AccountSettingsHandler extends MatrixClientBackedSettingsHa
             return !content['disable'];
         }
 
+        if (settingName === "notificationsEnabled") {
+            return this._getSettings("m.notification.sound");
```
Comment (turt2live) [inline src/settings/handlers/AccountSettingsHandler.js]: With my spec hat on: please don't use the m.* namespace without a proposal post-FCP. The rationale is that anything experimental should be in a custom namespace (either `im.vector` or `uk.half-shot` or something) to avoid potential clashes with future Matrix things. Riot in particular is a great test bed for these kinds of things because the settings system is capable of handling a migration period for if/when a proposal is approved and things move to m.* namespaces.

**Comment by Half-Shot** [comment]
Looks like a not-my-problem fail:

image

**Comment by dnaf** [comment]
> Removing a sound from a room doesn't remove the tag. 

A solution to this could be to give every room its own tag and change the `source`s whenever a rooms notification sound is changed. To me, this seems like a cleaner way to do things.

I don't think this is really an edge case; the odds of a user changing a notification sound for a room several times as they test different sounds and see which they'd prefer seems high.

**Comment by Half-Shot** [comment]
>A solution to this could be to give every room its own tag and change the sources whenever a rooms notification sound is changed. To me, this seems like a cleaner way to do things.

Honestly we could just keep a counter of which rooms use which <audio> tags and reference count it, and just remove them on 0. The reason I've not done this is that I'm not convinced that it's going to cause a great impact to Riot if you *do* have lots of audio tags open. It may leave a few extra KB in memory, but it would be cleared on next restart anyway.

**Comment by Half-Shot** [comment]
is it possible to get a quick review on this?

**Comment by jryans** [comment]
I have added it to the team queue.

File: src/i18n/strings/en_EN.json [hunk 3]
```diff
@@ -1618,5 +1618,9 @@
     "If you didn't remove the recovery method, an attacker may be trying to access your account. Change your account password and set a new recovery method immediately in Settings.": "If you didn't remove the recovery method, an attacker may be trying to access your account. Change your account password and set a new recovery method immediately in Settings.",
     "Failed to set direct chat tag": "Failed to set direct chat tag",
     "Failed to remove tag %(tagName)s from room": "Failed to remove tag %(tagName)s from room",
-    "Failed to add tag %(tagName)s to room": "Failed to add tag %(tagName)s to room"
+    "Failed to add tag %(tagName)s to room": "Failed to add tag %(tagName)s to room",
+    "Sounds": "Sounds",
+    "Custom Notification Sounds": "Notification sound",
+    "Set a new custom sound": "Set a new custom sound",
+    "Reset to default sound": "Reset to default sound"
```
Comment (turt2live) [inline src/i18n/strings/en_EN.json]: Please use the i18n script so these get sorted correctly into the file rather than manually inserting them

File: src/settings/Settings.js [hunk 4]
```diff
@@ -118,6 +119,9 @@ export const SETTINGS = {
         supportedLevels: LEVELS_FEATURE,
         default: false,
     },
+    "feature_notification_sounds": {
+        isFeature: true,
+        displayName: _td("Custom Notification Sounds"),
```
Comment (turt2live) [inline src/settings/Settings.js]: Syntax error: Missing closing curly brace

**Comment by Half-Shot** [comment]
Thanks for the comprehensive feedback @turt2live, will get to it :)

**Comment by Half-Shot** [comment]
@turt2live I believe I have done everything I can my side now.

**Comment by nadonomy** [comment]
@Half-Shot nice!

I know you're scratching a personal itch and this is in labs, but there's some 2 trivial UI changes that would be good to see:

1. We don't use black buttons in the UI. You're already decorating [Save] with `mx_AccessibleButton_kind_primary` and you'll want some visual distinction, with [Browse] carrying less weight, so I'd add a new CSS class for [Browse] with:

```
    color: $accent-color;
    background-color: transparent;
    border: 1px solid $accent-color;
```
We use similar styles in modals and elsewhere already.

2. I'd also decorate [Reset] with `mx_AccessibleButton_kind_danger` instead, and shorten 'Reset to default sound' to simply 'Reset'. You'll infer the function contextually, it commands less attention visually and the feature might benefit from internationalised strings.

**Comment by Half-Shot** [comment]
@nadonomy Thanks, these seem like good ideas. I'll go make the changes :)

**Comment by Half-Shot** [comment]
@nadonomy this is done

**Comment by nadonomy** [comment]
Great, thanks!

**Comment by Half-Shot** [comment]
Tagging @ara4n

**Comment by lampholder** [comment]
I've had a play and it LGTM :)

**Comment by Half-Shot** [comment]
Thanks :), does that mean I have met the criteria for labs now :innocent: ?

**Comment by jryans** [comment]
The remaining steps as I understand it are:

* Remove the labs / feature flag, as @lampholder feels it's okay to go in without that
* Request a fresh code review from riot-web team once that's done to double-check the code before landing

Almost there! 😁

File: src/Notifier.js [hunk 5]
```diff
@@ -100,10 +100,55 @@ const Notifier = {
...
+    _playAudioNotification: async function(ev, room) {
+        const sound = await this.getSoundForRoom(room.roomId);
+        console.log(`Got sound ${sound && sound.name || "default"} for ${room.roomId}`);
+
+        try {
+            const selector = document.querySelector(sound ? `audio[src='${sound.url}']` : "#messageAudio");
+            let audioElement = selector;
+            if (!selector) {
+                if (!sound) {
+                    console.error("Tried to play alert sound but missing #messageAudio");
```
Comment (turt2live) [inline src/Notifier.js]: this error could be clearer. Maybe this?
```suggestion
                    console.error("No audio element or sound to play for notification");
```

File: src/components/views/dialogs/RoomSettingsDialog.js [hunk 6]
```diff
@@ -67,6 +69,13 @@ export default class RoomSettingsDialog extends React.Component {
             "mx_RoomSettingsDialog_rolesIcon",
             <RolesRoomSettingsTab roomId={this.props.roomId} />,
         ));
+
```
Comment (turt2live) [inline src/components/views/dialogs/RoomSettingsDialog.js]: Extreme nit: no whitespace around the tab definition to match the others.

File: src/Notifier.js (hunk 5 above)
Comment (Half-Shot) [reply src/Notifier.js]: works for me

**Comment by Half-Shot** [comment]
@turt2live all good to merge now? :3

**Comment by turt2live** [comment]
giphy
